### ISSUE TITLE
Documentation/recipe for containerized builds and deployments

### ISSUE THREAD

**Issue by jeremydw** [grow/grow#402]
One of Grow's taglines is "develop everywhere, deploy anywhere". In practice, this means that development exists in various locations (workstations, VMs, GitHub, etc.) that have access to Git, and deployments exist anywhere a static build can be copied (e.g. Google Cloud Storage, internal/proprietary web server, Google App Engine, S3, etc.).

That said, in order to actually "develop everywhere and deploy anywhere", one must still install and run an instance of Grow (and, typically, npm, node, etc.) to build the site. Once a user has set up their environment once, this is usually OK - however, it's too much to expect non-technical users to be able to install and configure the development tool locally.

Moreover, even for developers, it can be annoying to execute a build manually or manage multiple versions of Grow especially when working on multiple projects.

This is where containerization can come in, and Google Cloud has a new feature (Container Builders - https://cloud.google.com/container-builder/docs/quickstart-gcloud) that will make this really nice and easy for an infrastructure-less system like Grow. Gcloud builders is like Travis or Circle as a service, and it's API-driven.

Here's how I envision it working:

1. Write documentation or develop a recipe (can be as simple as a Makefile or some other script, paired with a gcloud builders config file) that will use gcloud builders to build (and optionally, stage to webreview) a build. The build should be controlled by the script/config file we document.
2. Artifacts should be dumped to GCS, including the manifest files so that step (3) knows what to add/edit/delete.
3. There should be a separate script that can the artifacts from GCS and apply the diff to the deployment destination.

In other words, each time a Git commit is observed, gcloud builders kicks off, builds and writes the site to GCS for that build, and makes the artifacts available for deployment.

cc @drgrove @kezlya per discussion

**Comment by drGrove** [comment]
By "Git commit is observed" are you thinking:
- webhook integration?
- Post commit hook integration
- Both/either

**Comment by jeremydw** [comment]
Yeah, either. And for systems like Gerrit that might not support webhooks then polling or a cron could be used. For example - a micro App Engine app could be deployed or some Firebase function (for example) to detect whether a build has occurred for a commit yet, and if not, build it.

I completed a tiny bit of testing, and the following minimal `cloudbuild.yaml` file works great using Google Cloud Builders! A simple Grow site without too much fancy stuff takes just about 1.5 mins to build and stage. This includes the `npm install` step that's kicked off by `grow install`. Since there is no caching on Gcloud builders (unlike Travis or Circle CI), right now that install step runs each time.

We've used Circle CI for a bunch of projects and even without Circle's caching, gcloud builders is ~1 minute faster for the same build. (Probably because of how long Circle takes to restore a cache, collect artifacts, etc.)

```
steps:
- name: grow/growsdk
  args: ['install']
- name: grow/growsdk
  args: ['stage', '--api-key', ${_WEBREVIEW_API_KEY}]
```

**Comment by jeremydw** [comment]
Grow's Docker image on DockerHub (where `grow/growsdk` comes from in the `cloudbuild.yaml` example) – https://hub.docker.com/r/grow/growsdk/ – works well.

**Comment by Zoramite** [comment]
We now have some really good docker images that allow for direct usage of the `grow` command (such as in the container cloud) or in CI like CircleCI that I think makes this now complete.
